lib/SIL/Utils/OwnershipUtils.cpp

```diff
@@ -398,6 +412,67 @@ bool swift::findUsesOfSimpleValue(SILValue value,
   return true;
 }
 
+bool swift::visitGuaranteedForwardingPhisForSSAValue(
+    SILValue value, function_ref<bool(Operand *)> visitor) {
+  assert(isa<BeginBorrowInst>(value) || isa<LoadBorrowInst>(value) ||
+         (isa<SILPhiArgument>(value) &&
+          value->getOwnershipKind() == OwnershipKind::Guaranteed));
+  // guaranteedForwardingOps is a collection of all transitive
+  // GuaranteedForwarding uses of \p value. It is a set, to avoid repeated
+  // processing of structs and tuples which are GuaranteedForwarding.
+  SmallSetVector<Operand *, 4> guaranteedForwardingOps;
+  // Collect first-level GuaranteedForwarding uses, and call the visitor on any
+  // GuaranteedForwardingPhi uses.
+  for (auto *use : value->getUses()) {
+    if (use->getOperandOwnership() == OperandOwnership::GuaranteedForwarding) {
+      guaranteedForwardingOps.insert(use);
+      continue;
+    }
+    if (use->getOperandOwnership() ==
+        OperandOwnership::GuaranteedForwardingPhi) {
+      if (!visitor(use)) {
+        return false;
+      }
+      continue;
+    }
+  }
+
+  // Transitively, collect GuaranteedForwarding uses.
+  for (unsigned i = 0; i < guaranteedForwardingOps.size(); i++) {
+    // TODO: We are assuming the GuaranteedForwarding results are @guaranteed
+    // and can further forward @guaranteed values. This assumption exists in
+    // other places in SILOWnershipVerifier as well. We should add verification
+    // for this.
+    for (auto val : guaranteedForwardingOps[i]->getUser()->getResults()) {
+      for (auto *valUse : val->getUses()) {
+        if (valUse->getOperandOwnership() !=
+            OperandOwnership::GuaranteedForwarding) {
+          continue;
+        }
+        guaranteedForwardingOps.insert(valUse);
+      }
+    }
+  }
+
+  // Iterate guaranteedForwardingOps and call the visitor on
+  // GuaranteedForwardingPhi uses.
+  for (unsigned i = 0; i < guaranteedForwardingOps.size(); i++) {
+    for (auto val : guaranteedForwardingOps[i]->getUser()->getResults()) {
```

Comment:
Is this loop necessary? All guaranteed forwarding op results were handled in the loop above. Why not call visitor directly on any phi uses in the above loop?

Reply:
We need to process GuaranteedForwarding op results transitively, to find all GuaranteedForwardingPhi uses.
```
%b = begin_borrow ...
%ex1 = struct_extract %b ...
%ex2 = struct_extract %ex1 ...
....
br bbdest(%ex2)
```

We need the second loop to find `%ex2`

Reply:
Thanks for clarifying. I have fixed this.